File: kdf.py
#!/usr/bin/env python3
# coding=utf-8
"""
This is a simple example of doing an elliptic curve Diffie-Hellman ECDH) key exchange.

It allows two parties to jointly agree on a shared secret using an insecure channel.

NOTE: Cryptography version 2.0 in combination with very new versions of OpenSSL support a simpler
interface to use Curve25519 via from cryptography.hazmat.primitives.asymmetric.x25519 import X25519PrivateKey
"""
import sys

import colorama
from colorama import Fore
from cryptography.hazmat.backends import default_backend
from cryptography.hazmat.primitives import hashes
from cryptography.hazmat.primitives.kdf.kbkdf import CounterLocation, KBKDFHMAC, Mode
from cryptography.hazmat.primitives.kdf.pbkdf2 import PBKDF2HMAC


# Choose a key derivation function (KDF)
PBKDF = 'password-based'
KBKDF = 'key-based'
KDF = PBKDF

USAGE = """Usage:
    {0} <password> <salt>

Example:
    {0} 1337P@ssw0rd hex:0123456789ABCDEF0123456789ABCDEF0123456789ABCDEF0123456789ABCDEF"
""".format(sys.argv[0])


def print_usage() -> None:
    """Display intended command-line usage"""
    print(Fore.CYAN + USAGE)


if __name__ == '__main__':
    colorama.init(autoreset=True)

    # Parse the command-line arguments
    expected_args = 2
    if len(sys.argv) != expected_args + 1:
        print(Fore.RED + 'Expected {} arguments, but got {}'.format(expected_args, len(sys.argv) - 1))
        print_usage()
        sys.exit(1)

    # Read the password
    password = sys.argv[1].encode()

    # Read the salt from command line as hex digits (all caps) prefaced by "hex:"
    if sys.argv[2].startswith('hex:'):
        salt = bytes.fromhex(sys.argv[2][4:])
    else:
        print(Fore.RED + 'invalid salt format')
        sys.exit(2)

    # cryptography backend (i.e. OpenSSL)
    backend = default_backend()

    # instance of a HashAlgorithm to use with the kdf
    hash_algo = hashes.SHA512()

    print(Fore.CYAN + 'Using {} Key Derivation Function (KDF) with {} hash algorithm \n'.format(KDF, hash_algo.name))
    print(Fore.LIGHTMAGENTA_EX + 'Password: {!r}'.format(password.decode()))
    print(Fore.LIGHTBLUE_EX + 'Salt: {!r}'.format(salt.hex().upper()))

    # Derive a key using a Key Derivation Function (KDF)
    if KDF == PBKDF:
        # The PBKDF2 (Password Based Key Derivation Function 2) is part of RSA Laboratories' Public-Key Cryptography
        # Standards (PKCS) series, specifically PKCS #5 v2.0, also published as Internet Engineering Task Force's
        # RFC 2898.  It is available in most cryptographic libraries.

        # Salts should be randomly generated
        kdf = PBKDF2HMAC(algorithm=hash_algo,   # An instance of HashAlgorithm
                         length=32,             # The desired length of the derived key in bytes
                         salt=salt,             # A salt. Secure values are 128-bits (16 bytes) or longer
                         iterations=100000,     # The number of iterations to perform of the hash function
                         backend=backend)       # An instance of PBKDF2HMACBackend
    else:

        # The KBKDF (Key Based Key Derivation Function) used here is defined by the NIST SP 800-108 document, to be used
        # to derive additional keys from a key that has been established through an automated key-establishment scheme
        label = b'KBKDF HMAC Label'
        context = b'KBKDF HMAC Context'
        kdf = KBKDFHMAC(algorithm=hash_algo,    # An instance of HashAlgorithm
                        mode=Mode.CounterMode,  # The desired mode of the PRF.A value from the Mode enum
                        length=32,              # The desired length of the derived key in bytes
                        rlen=4,                 # An integer that indicates the length of the counter in bytes
                        llen=4,                 # An integer that indicates the length of the length in bytes
                        location=CounterLocation.BeforeFixed,   # The desired location of the counter
                        label=label,            # Application specific label information.  Byte string or None
                        context=context,        # Application specific context information.  Byte string or None
                        fixed=None,             # Instead of specifying label and context, supply your own fixed data
                        backend=backend)        # A cryptography backend HashBackend instance

    # Derive a new key from the shared secret
    derived_key = kdf.derive(password)

    # Make sure shared keys agree
    print(Fore.GREEN + '\nDerived AES key ({} bits): {}'.format(len(derived_key) * 8, derived_key.hex().upper()))

File: kdf.c
/*
 *  Example Key Derivation Function (KDF) program using mbedTLS
 *
 *  This file uses a password-based key derivation function specified in PKCS#5 PBKDF2 and implemented in mbedTLS in the
 * mbedtls_pkcs5_pbkdf2_hmac() function in the pkcs5.h/.c files.
 *
 * PBDKF2 is a KDF with sliding computation cost aimed to reduce the vulnerability of encrypted keys to brute force
 * attacks.  PBKDF2 is part of RSA Laboratorie's Public-Key Cryptography Standards (PKCS) series, specifically PKCS#5
 * v2.0, also published as Internet Engineering Task Force's RFC 2898.  It supercedes PBKDF1, which could only produce
 * keys up to 160 bits long.
 *
 * There are better KDF functions available which address weaknesses in PBDKF2, but PBKDF2 is widely available in most
 * libraries.
 *
 * PBKDF2 applies a pseudorandom function, such as a hash-based message authentication code (HMAC), to the input
 * password along with a salt value and repeats the process many times to produce a derived key, which can then be used
 * as a cryptographic key in subsequent operations. The added computation work makes password cracking much more
 * difficult, and is known as key stretching.
 */

#if !defined(MBEDTLS_CONFIG_FILE)
#include "mbedtls/config.h"
#else
#include MBEDTLS_CONFIG_FILE
#endif

#if defined(MBEDTLS_PLATFORM_C)
#include "mbedtls/platform.h"
#else
#include <stdio.h>
#define mbedtls_printf  printf
#endif

#include <string.h>

#if !defined(MBEDTLS_SHA512_C) || !defined(MBEDTLS_MD_C)
int main( void )
{
    mbedtls_printf( "MBEDTLS_SHA512_C and/or MBEDTLS_MD_C not defined\n");
    return( 0 );
}
#else

#include "mbedtls/entropy.h"    // entropy gathering
#include "mbedtls/ctr_drbg.h"   // CSPRNG
#include "mbedtls/error.h"      // error code to string conversion
#include "mbedtls/md.h"         // hash algorithms
#include "mbedtls/pkcs5.h"      // pbkdf2 KDF

#define USAGE   \
    "\n  %s <password>\n" \
    "\n  example: %s 1337P@ssw0rd\n" \
    "\n", argv[0], argv[0]

// Size of buffer used to translate mbed TLS error codes into a string representation
#define MBED_ERR_BUF 80

// Define which hash algorithm to use for the HMAC
//#define HASH_ALGORITHM MBEDTLS_MD_SHA256
//#define HASH_ALGORITHM MBEDTLS_MD_SHA384
#define HASH_ALGORITHM MBEDTLS_MD_SHA512

// Length of salt
#define SALT_BYTES 16

// Iteration count (how many times to iteratively run through the HMAC)
#define ITERATION_COUNT 100000

// Length of generated key in bytes
#define KEY_BYTES 32


int main( int argc, char *argv[] )
{
    int ret;
    mbedtls_entropy_context entropy;
    mbedtls_ctr_drbg_context ctr_drbg;
    mbedtls_md_context_t sha_ctx;

    // Create buffers for the salt, output key, and string representation of error codes
    unsigned char salt[SALT_BYTES];
    unsigned char key[KEY_BYTES];
    char mbed_err[MBED_ERR_BUF];

    // Process command-line arguments - require password passed as 1st argument on command line
    if(argc < 2)
    {
        mbedtls_printf( USAGE );
        return -1;
    }

    // Password to use when generating the derived key
    unsigned char *password = (unsigned char*) argv[1];

    // Length of password
    size_t plen = strlen(argv[1]);

    mbedtls_printf( "Password: '%s' (length %zu)\n", argv[1], plen);

    mbedtls_ctr_drbg_init( &ctr_drbg );
    mbedtls_entropy_init( &entropy );
    mbedtls_md_init( &sha_ctx );

    mbedtls_printf( "Using Hash Algorithm:  ");
    switch(HASH_ALGORITHM)
    {
    case MBEDTLS_MD_SHA256:
        mbedtls_printf( "SHA256\n" );
        break;
    case MBEDTLS_MD_SHA384:
        mbedtls_printf( "SHA384\n" );
        break;
    case MBEDTLS_MD_SHA512:
        mbedtls_printf( "SHA512\n" );
        break;
    default:
        mbedtls_printf( "ERROR - Invalid Hash Algorithm selected!\n" );
        goto exit;
    }

    // Initialize random number generation
    mbedtls_printf( "  . Seeding the random number generator..." );
    fflush( stdout );
    const char pers[] = "kdf";
    ret = mbedtls_ctr_drbg_seed( &ctr_drbg, mbedtls_entropy_func, &entropy, (const unsigned char *) pers, sizeof pers );
    if( ret != 0 )
    {
        mbedtls_printf( " failed\n  ! mbedtls_ctr_drbg_seed returned %d\n", ret );
        goto exit;
    }
    mbedtls_printf( " ok\n" );


    // Select hash algorithm to use for the HMAC and allocate internal structures
    mbedtls_printf( "  . Setting up hash algorithm context..." );
    fflush( stdout );
    ret = mbedtls_md_setup( &sha_ctx,   // MD context to set up
                            mbedtls_md_info_from_type( HASH_ALGORITHM ),    // md_info struct MD to use
                            1 );    // non-zero implies HMAC is going to be used (0 saves memory, but is less secure)
    if( ret != 0 )
    {
        mbedtls_printf( "  ! mbedtls_md_setup() returned -0x%04x\n", -ret );
        goto exit;
    }
    mbedtls_printf( " ok\n" );

    // Generate random data for the salt
    mbedtls_printf( "  . Generating a random salt..." );
    ret = mbedtls_ctr_drbg_random( &ctr_drbg, salt, SALT_BYTES );
    if( ret != 0 )
    {
        printf( "mbedtls_ctr_drbg_random failed to extract key - returned -0x%04x\n", -ret );
        goto exit;
    }
    mbedtls_printf( " ok\n" );

    // Derive a key from a password using PBKDF2 function with HMAC
    mbedtls_printf( "  . Deriving a key from the password using PBKDF2 with HMAC..." );
    ret = mbedtls_pkcs5_pbkdf2_hmac( &sha_ctx,  // Generic HMAC context
                                     password,  // Password to use when generating key
                                     plen,      // Length of password
                                     salt,      // Salt to use when generating key
                                     SALT_BYTES,// Length of salt
                                     ITERATION_COUNT, //Iteration count
                                     KEY_BYTES, // Length of generated key in bytes
                                     key );     // Generated key. Must be at least as big as previous argument
    if( ret != 0 )
    {
        mbedtls_printf( " failed\n  ! mbedtls_pkcs5_pbkdf2_hmac - returned -0x%04x\n", -ret );
        goto exit;
    }
    mbedtls_printf( " ok\n" );

    mbedtls_printf( "Random %d byte salt (hex):  ", SALT_BYTES );
    for(size_t i = 0; i < SALT_BYTES; i++)
    {
        mbedtls_printf("%02X", salt[i]);
    }
    printf("\n");

    mbedtls_printf( "Derived a %d byte key (hex):  ", KEY_BYTES );
    for(size_t i = 0; i < KEY_BYTES; i++)
    {
        mbedtls_printf("%02X", key[i]);
    }
    printf("\n");


    // NOTE:  The salt affects the derived key, so it needs to be shared, though it can be sent in the clear

exit:
    // If there was an error, translate an mbed TLS error code into a string representation
    if( ret != 0 )
    {
        mbedtls_strerror(ret, mbed_err, MBED_ERR_BUF);
        mbedtls_printf( "mbedTLS ERROR: %s\n", mbed_err);
    }

#if defined(_WIN32)
    mbedtls_printf( "  + Press Enter to exit this program.\n" );
    fflush( stdout ); getchar();
#endif

    mbedtls_ctr_drbg_free( &ctr_drbg );
    mbedtls_entropy_free( &entropy );
    mbedtls_md_free( &sha_ctx );

    return( ret != 0 );
}
#endif /* MBEDTLS_MD_SHA512 && MBEDTLS_MD_C */
